added support for saml login

kevinpapst opened the pull request:
## Description

**Only tested with Google apps SAML implementation**

In order to test this branch, do a checkout and then:
- composer install
- edit `config/packages/local.yaml` with the adjusted contents below
- `bin/console cache:clear --env=prod`
- `bin/console cache:warmup --env=prod`
- Test the new login button ...

**YOU FIND THE REQUIRED CONFIGURATION IN THE DOCUMENTATION**
See https://www.kimai.org/documentation/saml.html

Fixes #1227 

Links:
- [Dynatrace: Setup groups in G Suite](https://www.dynatrace.com/support/help/how-to-use-dynatrace/user-management-and-sso/manage-users-and-groups-with-saml/saml-gsuite/#preparing-group-mapping)
- [Google: Creating custom attributes using the user schema](https://support.google.com/cloudidentity/answer/6327792?hl=en&ref_topic=7558947)

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
- [x] I verified that my code applies to the guidelines (`composer kimai:code-check`)
- [ ] I updated the documentation (see [here](https://github.com/kimai/www.kimai.org/tree/master/_documentation))
- [x] I agree that this code is used in Kimai and will be published under the [MIT license](https://github.com/kevinpapst/kimai2/blob/master/LICENSE)

timlegge commented:
Hi

I get an error on successful saml response:

bash-5.0$ tail -f var/log/prod.log 
[2020-01-24 04:02:57] security.DEBUG: Checking for guard authentication credentials. {"firewall_key":"secured_area","authenticators":1} []
[2020-01-24 04:02:57] security.DEBUG: Checking support on guard authenticator. {"firewall_key":"secured_area","authenticator":"App\\Security\\TokenAuthenticator"} []
[2020-01-24 04:02:57] security.DEBUG: Guard authenticator does not support the request. {"firewall_key":"secured_area","authenticator":"App\\Security\\TokenAuthenticator"} []
[2020-01-24 04:02:57] request.CRITICAL: Uncaught PHP Exception Doctrine\DBAL\Exception\NotNullConstraintViolationException: "An exception occurred while executing 'INSERT INTO kimai2_users (username, username_canonical, email, email_canonical, enabled, salt, password, last_login, confirmation_token, password_requested_at, roles, alias, registration_date, title, avatar, api_token) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [null, null, null, null, 1, null, "", null, null, null, "a:0:{}", null, "2020-01-24 04:02:57", null, null, null]:  SQLSTATE[23000]: Integrity constraint violation: 19 NOT NULL constraint failed: kimai2_users.username" at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractSQLiteDriver.php line 39 {"exception":"[object] (Doctrine\\DBAL\\Exception\\NotNullConstraintViolationException(code: 0): An exception occurred while executing 'INSERT INTO kimai2_users (username, username_canonical, email, email_canonical, enabled, salt, password, last_login, confirmation_token, password_requested_at, roles, alias, registration_date, title, avatar, api_token) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [null, null, null, null, 1, null, \"\", null, null, null, \"a:0:{}\", null, \"2020-01-24 04:02:57\", null, null, null]:\n\nSQLSTATE[23000]: Integrity constraint violation: 19 NOT NULL constraint failed: kimai2_users.username at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractSQLiteDriver.php:39, Doctrine\\DBAL\\Driver\\PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 19 NOT NULL constraint failed: kimai2_users.username at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:123, PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 19 NOT NULL constraint failed: kimai2_users.username at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:121)"} []
[2020-01-24 04:10:22] console.ERROR: Error thrown while running command "'kimai:create-user' superadmin 'admin@kimai.local' ROLE_SUPER_ADMIN changemeplease". Message: "The Translator does not support the following options: 'cache_vary'." {"exception":"[object] (Symfony\\Component\\Translation\\Exception\\InvalidArgumentException(code: 0): The Translator does not support the following options: 'cache_vary'. at /opt/kimai/vendor/symfony/framework-bundle/Translation/Translator.php:77)","command":"'kimai:create-user' superadmin 'admin@kimai.local' ROLE_SUPER_ADMIN changemeplease","message":"The Translator does not support the following options: 'cache_vary'."} []
[2020-01-24 04:13:42] request.INFO: Matched route "saml_acs". {"route":"saml_acs","route_parameters":{"_route":"saml_acs","_controller":"App\\Auth\\Controller\\SamlController::assertionConsumerServiceAction"},"request_uri":"https://kimai-saml.alc.ca/auth/saml/acs","method":"POST"} []
[2020-01-24 04:13:42] security.DEBUG: Checking for guard authentication credentials. {"firewall_key":"secured_area","authenticators":1} []
[2020-01-24 04:13:42] security.DEBUG: Checking support on guard authenticator. {"firewall_key":"secured_area","authenticator":"App\\Security\\TokenAuthenticator"} []
[2020-01-24 04:13:42] security.DEBUG: Guard authenticator does not support the request. {"firewall_key":"secured_area","authenticator":"App\\Security\\TokenAuthenticator"} []
[2020-01-24 04:13:43] request.CRITICAL: Uncaught PHP Exception Doctrine\DBAL\Exception\NotNullConstraintViolationException: "An exception occurred while executing 'INSERT INTO kimai2_users (username, username_canonical, email, email_canonical, enabled, salt, password, last_login, confirmation_token, password_requested_at, roles, alias, registration_date, title, avatar, api_token) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [null, null, null, null, 1, null, "", null, null, null, "a:0:{}", null, "2020-01-24 04:13:42", null, null, null]:  SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'username' cannot be null" at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php line 103 {"exception":"[object] (Doctrine\\DBAL\\Exception\\NotNullConstraintViolationException(code: 0): An exception occurred while executing 'INSERT INTO kimai2_users (username, username_canonical, email, email_canonical, enabled, salt, password, last_login, confirmation_token, password_requested_at, roles, alias, registration_date, title, avatar, api_token) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [null, null, null, null, 1, null, \"\", null, null, null, \"a:0:{}\", null, \"2020-01-24 04:13:42\", null, null, null]:\n\nSQLSTATE[23000]: Integrity constraint violation: 1048 Column 'username' cannot be null at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php:103, Doctrine\\DBAL\\Driver\\PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'username' cannot be null at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:123, PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'username' cannot be null at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:121)"} []
^C

Possible I am doing something wrong...

timlegge commented:
Ah.  It requires me to create the username first.  It should default to creating the user like ldap.

kevinpapst commented:
Accepted: `email` is now the only required field, `username` will fallback to `email` if not set.
Did you try groups already?

timlegge commented:
not yet.  home to today

On Fri., Jan. 24, 2020, 7:21 a.m. Kevin Papst, <notifications@github.com>
wrote:

> Accepted: email is now the only required field, username will fallback to
> email if not set.
> Did you try groups already?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kevinpapst/kimai2/pull/1408?email_source=notifications&email_token=AAH3N6YUCPQJ5BEVZSQWSFDQ7LFM5A5CNFSM4KKZAKR2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJ2PSTI#issuecomment-578091341>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH3N66FN46L247IVNKKEEDQ7LFM5ANCNFSM4KKZAKRQ>
> .
>

timlegge commented:
Hi

I successfully tested a login with a Microsoft Azure IDp however upon the second login it attempts to create a duplicate account.  

[2020-01-24 14:25:06] request.CRITICAL: Uncaught PHP Exception Doctrine\DBAL\Exception\UniqueConstraintViolationException: "An exception occurred while executing 'INSERT INTO kimai2_users (username, username_canonical, email, email_canonical, enabled, salt, password, last_login, confirmation_token, password_requested_at, roles, alias, registration_date, title, avatar, api_token) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params ["tl@example.com", "tl@example.com", "tl@example.com", "tl@example.com", 1, null, "", null, null, null, "a:0:{}", null, "2020-01-24 14:25:06", null, null, null]:  SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'tl@example.com' for key 'UNIQ_B9AC5BCE92FC23A8'" at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php line 55 {"exception":"[object] (Doctrine\\DBAL\\Exception\\UniqueConstraintViolationException(code: 0): An exception occurred while executing 'INSERT INTO kimai2_users (username, username_canonical, email, email_canonical, enabled, salt, password, last_login, confirmation_token, password_requested_at, roles, alias, registration_date, title, avatar, api_token) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [\"tl@example.com\", \"tl@example.com\", \"tl@example.com\", \"tl@example.com\", 1, null, \"\", null, null, null, \"a:0:{}\", null, \"2020-01-24 14:25:06\", null, null, null]:\n\nSQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'tl@example.com' for key 'UNIQ_B9AC5BCE92FC23A8' at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php:55, Doctrine\\DBAL\\Driver\\PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'tl@example.com' for key 'UNIQ_B9AC5BCE92FC23A8' at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:123, PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'tl@example.com' for key 'UNIQ_B9AC5BCE92FC23A8' at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:121)"} []

kevinpapst commented:
Maybe because of your previous tests? Check your `kimai2_users` database existing accounts. The unique index is on the `username_canonical` column.

timlegge commented:
No, there were only 3 users;  My superadmin, google account, and the microsoft account

I deleted the microsoft account and was able to log in.  I closed the browser and tried again and I was not able to login

kevinpapst commented:
Please add a `var_dump` in `retrieveUser` the file: vendor/kevinpapst/oneloginsaml-bundle/Security/Authentication/Provider/SamlProvider.php in line 73, so the message looks like this:

```
        try {
            var_dump($token);exit;
            return $this->userProvider->loadUserByUsername($token->getUsername());
        } catch (UsernameNotFoundException $e) {
            if ($this->userFactory instanceof SamlUserFactoryInterface) {
                return $this->generateUser($token);
            }
            
            throw $e;
        }
``` 

and let me know what you see. Maybe the user has the wrong email?

timlegge commented:
Sorted the duplicate I changed the local.yaml mapping from 

username: $Email 

to

username: $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name

How do the groups mapping work?  An can I map the full name Full Name?

kevinpapst commented:
I think you should be able to remove the username from the mapping completely, as it will default to that field - but I am not entirely sure.

You can try to set the `alias` for the full name.

timlegge commented:
Working on the groups.  I have
        groups:
            attribute: groups
            mapping:
                Kimai - Administrator: ROLE_ADMIN
                Kimai - Team Lead: ROLE_TEAMLEAD

Which as you may notice have spaces :-)

when  /opt/kimai/src/Auth/User/SamlUserFactory.php reads them ( var_dump($this->groupMapping); at line 52

I get:
array(2) { ["Kimai _ Administrator"]=> string(10) "ROLE_ADMIN" ["Kimai _ Team Lead"]=> string(13) "ROLE_TEAMLEAD" }

For some reason it is converting the - to an _

Any ideas?

kevinpapst commented:
Oh please not 😬 okay right on, I will convert that "array like" syntax to the same that is used for LDAP. Now I can remember the reason ^^

kevinpapst commented:
@timlegge I updated the configuration syntax, you need to change the block in `kimai`.

The attribute mapping from:
```
        mapping:
            email: $Email
            username: $Email
            alias: $FullName
```
to
```
        mapping:
            - { saml: $Email, kimai: email }
            - { saml: $Email, kimai: username }
            - { saml: $FullName, kimai: alias }
```

Then `groups` was renamed to `roles` and the array syntax under `roles` was also changed like the mapping above.

See https://www.kimai.org/documentation/saml.html

Full example:
```
kimai:
    saml:
        activate: true
        title: Login with Google
        mapping:
            - { saml: $Email, kimai: email }
            - { saml: $Email, kimai: username }
            - { saml: $FullName, kimai: alias }
        roles:
            attribute: Roles
            mapping:
                - { saml: Admins, kimai: ROLE_ADMIN }
                - { saml: Management, kimai: ROLE_TEAMLEAD }
```

timlegge commented:
testing for the last bit...

On Fri., Jan. 24, 2020, 6:14 p.m. Kevin Papst, <notifications@github.com>
wrote:

> @timlegge <https://github.com/timlegge> I updated the configuration
> syntax, you need to change the block in kimai.
>
> The attribute mapping from:
>
>         mapping:
>             email: $Email
>             username: $Email
>             alias: $FullName
>
> to
>
>         mapping:
>             - { saml: $Email, kimai: email }
>             - { saml: $Email, kimai: username }
>             - { saml: $FullName, kimai: alias }
>
> Then groups was renamed to roles and the array syntax under roles was
> also changed like the mapping above.
>
> See https://www.kimai.org/documentation/saml.html
>
> Full example:
>
> kimai:
>     saml:
>         activate: true
>         title: Login with Google
>         mapping:
>             - { saml: $Email, kimai: email }
>             - { saml: $Email, kimai: username }
>             - { saml: $FullName, kimai: alias }
>         roles:
>             attribute: Roles
>             mapping:
>                 - { saml: Admins, kimai: ROLE_ADMIN }
>                 - { saml: Management, kimai: ROLE_TEAMLEAD }
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kevinpapst/kimai2/pull/1408?email_source=notifications&email_token=AAH3N63TTA5K5DNUMWWDFIDQ7NR5XA5CNFSM4KKZAKR2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJ4IGUA#issuecomment-578323280>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH3N6ZQCFD6CCV2T3HK7JDQ7NR5XANCNFSM4KKZAKRQ>
> .
>

timlegge commented:
@kevinpapst Initial testing looks promising.  Nice!  

I have tested with one user (static group) against a Microsoft Azure Saml Identity Provider.

With the change to the ldap style arrays my groups complete with spaces and dashes (I know, I know) works and the user is created.

I intercepted the response from the server and tampered with the xml to attempt to grant extra permissions and as expected OneLogin caught the tampering with the signed Message - great

I have not changed the group that the user is in yet but I did set the roles in the user record to ''.  As I was logged in that changed my user to a normal user on page refresh

I tried to log in again and the permissions were that of a normal user.  It would be expected to refresh the user data on login.  Especially the title, roles and possibly alias

So far that is the only issue I have found.

kevinpapst commented:
> It would be expected to refresh the user data on login. Especially the title, roles and possibly alias

Things like that should be discussed upfront. The necessary change was really big, but it works now.

When updating, you need to remove these two lines from the config (see updates docs):
```
                user_factory: App\Auth\User\SamlUserFactory
                persist_user: true
```

timlegge commented:
Hi

Just testing this morning and saw the following error.  Realized that the bin/console kimai:update need to be run.

However now the saml process runs.  Request to IdP, response valid, I see: An authentication exception occurred. on the login page and no exceptions in the logs.

I checked the user table.  

roles:  a:1:{i:0;s:10:"ROLE_ADMIN";
auth: saml

Tim

[2020-01-27 12:40:29] request.INFO: Matched route "saml_acs". {"route":"saml_acs","route_parameters":{"_route":"saml_acs","_controller":"App\\Auth\\Controller\\SamlController::assertionConsumerServiceAction"},"request_uri":"https://kimai-saml.example.com/auth/saml/acs","method":"POST"} []
[2020-01-27 12:40:29] security.DEBUG: Checking for guard authentication credentials. {"firewall_key":"secured_area","authenticators":1} []
[2020-01-27 12:40:29] security.DEBUG: Checking support on guard authenticator. {"firewall_key":"secured_area","authenticator":"App\\Security\\TokenAuthenticator"} []
[2020-01-27 12:40:29] security.DEBUG: Guard authenticator does not support the request. {"firewall_key":"secured_area","authenticator":"App\\Security\\TokenAuthenticator"} []
[2020-01-27 12:40:29] request.CRITICAL: Uncaught PHP Exception Doctrine\DBAL\Exception\InvalidFieldNameException: "An exception occurred while executing 'SELECT t0.username AS username_1, t0.username_canonical AS username_canonical_2, t0.email AS email_3, t0.email_canonical AS email_canonical_4, t0.enabled AS enabled_5, t0.salt AS salt_6, t0.password AS password_7, t0.last_login AS last_login_8, t0.confirmation_token AS confirmation_token_9, t0.password_requested_at AS password_requested_at_10, t0.roles AS roles_11, t0.id AS id_12, t0.alias AS alias_13, t0.registration_date AS registration_date_14, t0.title AS title_15, t0.avatar AS avatar_16, t0.api_token AS api_token_17, t0.auth AS auth_18 FROM kimai2_users t0 WHERE t0.email_canonical = ? LIMIT 1' with params ["tim@example.com"]:  SQLSTATE[42S22]: Column not found: 1054 Unknown column 't0.auth' in 'field list'" at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php line 60 {"exception":"[object] (Doctrine\\DBAL\\Exception\\InvalidFieldNameException(code: 0): An exception occurred while executing 'SELECT t0.username AS username_1, t0.username_canonical AS username_canonical_2, t0.email AS email_3, t0.email_canonical AS email_canonical_4, t0.enabled AS enabled_5, t0.salt AS salt_6, t0.password AS password_7, t0.last_login AS last_login_8, t0.confirmation_token AS confirmation_token_9, t0.password_requested_at AS password_requested_at_10, t0.roles AS roles_11, t0.id AS id_12, t0.alias AS alias_13, t0.registration_date AS registration_date_14, t0.title AS title_15, t0.avatar AS avatar_16, t0.api_token AS api_token_17, t0.auth AS auth_18 FROM kimai2_users t0 WHERE t0.email_canonical = ? LIMIT 1' with params [\"tim@example.com\"]:\n\nSQLSTATE[42S22]: Column not found: 1054 Unknown column 't0.auth' in 'field list' at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php:60, Doctrine\\DBAL\\Driver\\PDOException(code: 42S22): SQLSTATE[42S22]: Column not found: 1054 Unknown column 't0.auth' in 'field list' at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:123, PDOException(code: 42S22): SQLSTATE[42S22]: Column not found: 1054 Unknown column 't0.auth' in 'field list' at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:121)"} []

kevinpapst commented:
Oh right, totally forgot to mention the update ^^

You need to either remove your previously added saml users or update the "auth" column with the value "saml" for them. I'd say: delete them and then retry.

timlegge commented:
auth = saml is in that user's record.  But I will delete and try again

timlegge commented:
Hi

I deleted the user.  I was able to login once after the initial login.

The user looks fine it has the role set to a:1:{i:0;s:10:"ROLE_ADMIN";} and auth set to saml

Tim

kevinpapst commented:
And after that initial login it fails?

timlegge commented:
yes.  :-) just updated the previous reply

kevinpapst commented:
And there is no log entry at all?
It works here without problems (G Suite) and I have no idea what the problem could be.
Can you switch to APP_ENV=dev ? You need to reinstall composer WITHOUT --no-dev.

timlegge commented:
Hi

I did:
composer install
export APP_ENV=dev
bin/console --env-dev cache:clear

Still no logs

I do see the correct session under var/sessions:

_sf2_attributes|a:2:{s:34:"_security.secured_area.target_path";s:37:"https://kimai-saml.example.com/en/homepage";s:24:"_csrf/https-authenticate";s:43:"TfACSEls59BlDqSOjbjFB3vnz9P5I_ZMFmSzcTBIqv0";}_sf2_meta|a:3:{s:1:"u";i:1580133681;s:1:"c";i:1580133188;s:1:"l";s:1:"0";}

The saml request and response was fine and the request was for the authenticated page.

timlegge commented:
Actually after the post of the saml response to auth/saml/acs the response is to redirect to the login page for some reason

timlegge commented:
In  src/Auth/Provider/SamlProvider.php 

I added:  var_dump($token);exit at line 74

The token looks fine but the $user = $this->userProvider->loadUserByUsername($token->getUsername()); at line 67 is null

timlegge commented:
Hi

In src/Auth/Provider/SamlProvider.php line 104 var_dump($user);exit; shows in the browser 

If I comment out 

$this->entityManager->flush();

Login works

Tim

kevinpapst commented:
Huh? Thats weird. Can you re-add the flush and dump the exception in the catch block?

timlegge commented:
Actually it shows an 

Something is wrong

A critical error occurred, please try again. You may have found a software problem, please contact your administrator if the problem persists. 

Now the logs show:

[2020-01-27 14:35:57] security.INFO: User has been authenticated successfully. {"username":"tim@example.com"} []
[2020-01-27 14:35:57] request.CRITICAL: Uncaught PHP Exception Doctrine\DBAL\Exception\UniqueConstraintViolationException: "An exception occurred while executing 'INSERT INTO kimai2_users (username, username_canonical, email, email_canonical, enabled, salt, password, last_login, confirmation_token, password_requested_at, roles, alias, registration_date, title, avatar, api_token, auth) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params ["tim@example.com", "tim@example.com", "timothy.legge@example.com", "timothy.legge@example.com", 1, null, "", "2020-01-27 14:35:57", null, null, "a:1:{i:0;s:10:\"ROLE_ADMIN\";}", "Timothy Legge", "2020-01-27 14:35:57", "My title", null, null, "saml"]:  SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'tim@example.com' for key 'UNIQ_B9AC5BCE92FC23A8'" at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php line 55 {"exception":"[object] (Doctrine\\DBAL\\Exception\\UniqueConstraintViolationException(code: 0): An exception occurred while executing 'INSERT INTO kimai2_users (username, username_canonical, email, email_canonical, enabled, salt, password, last_login, confirmation_token, password_requested_at, roles, alias, registration_date, title, avatar, api_token, auth) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [\"tim@example.com\", \"tim@example.com\", \"timothy.legge@example.com\", \"timothy.legge@example.com\", 1, null, \"\", \"2020-01-27 14:35:57\", null, null, \"a:1:{i:0;s:10:\\\"ROLE_ADMIN\\\";}\", \"Timothy Legge\", \"2020-01-27 14:35:57\", \"My Title\", null, null, \"saml\"]:\n\nSQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'tim@example.com' for key 'UNIQ_B9AC5BCE92FC23A8' at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php:55, Doctrine\\DBAL\\Driver\\PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'tim@example.com' for key 'UNIQ_B9AC5BCE92FC23A8' at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:123, PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'tim@example.com' for key 'UNIQ_B9AC5BCE92FC23A8' at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:121)"} []
[2020-01-27 14:35:57] security.DEBUG: Stored the security token in the session. {"key":"_security_secured_area"} []
tileg

kevinpapst commented:
Hm, that is again the username_canonical column.
Did you change your mapping config?

kevinpapst commented:
Maybe there is a mapping problem, can you try to move in file src/Auth/User/SamlUserFactory.php the code block (line 82)
```
        if (empty($user->getUsername())) {
            $user->setUsername($user->getEmail());
        }
```
to the end of the createUser() function in line 45?

timlegge commented:
I fixed the mapping issue - you mentioned that I might not need it so I removed it which caused the exception.  Whats the best way to dump that exception?

timlegge commented:
moving the code bloc had no impact

kevinpapst commented:
> I fixed the mapping issue - you mentioned that I might not need it so I removed it which caused the exception

I don't understand that.

The exception is already in the log, but using `var_dump($ex);exit;` should be enough

timlegge commented:
Sorry, typing is hard :-)

What I meant is that I likely caused my issue by removing the username mapping from local.yaml based on your previous comment that I might not need the mapping.  In case the mapping was causinig my issue.

I have since commented out the try and the exception was logged:

I appears to be attempting to insert the record for the user again:

[2020-01-27 14:50:30] request.CRITICAL: Uncaught PHP Exception Doctrine\DBAL\Exception\UniqueConstraintViolationException: "An exception occurred while executing 'INSERT INTO kimai2_users (username, username_canonical, email, email_canonical, enabled, salt, password, last_login, confirmation_token, password_requested_at, roles, alias, registration_date, title, avatar, api_token, auth) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params ["tim@example.com", "tim@example.com", "timothy.legge@example.com", "timothy.legge@example.com", 1, null, "", null, null, null, "a:1:{i:0;s:10:\"ROLE_ADMIN\";}", "Timothy Legge", "2020-01-27 14:50:30", "My Title", null, null, "saml"]:  SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'tim@example.com' for key 'UNIQ_B9AC5BCE92FC23A8'" at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php line 55 {"exception":"[object] (Doctrine\\DBAL\\Exception\\UniqueConstraintViolationException(code: 0): An exception occurred while executing 'INSERT INTO kimai2_users (username, username_canonical, email, email_canonical, enabled, salt, password, last_login, confirmation_token, password_requested_at, roles, alias, registration_date, title, avatar, api_token, auth) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [\"tim@example.com\", \"tim@example.com\", \"timothy.legge@example.com\", \"timothy.legge@example.com\", 1, null, \"\", null, null, null, \"a:1:{i:0;s:10:\\\"ROLE_ADMIN\\\";}\", \"Timothy Legge\", \"2020-01-27 14:50:30\", \"My Title\", null, null, \"saml\"]:\n\nSQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'tim@example.com' for key 'UNIQ_B9AC5BCE92FC23A8' at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php:55, Doctrine\\DBAL\\Driver\\PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'tim@example.com' for key 'UNIQ_B9AC5BCE92FC23A8' at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:123, PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'tim@example.com' for key 'UNIQ_B9AC5BCE92FC23A8' at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:121)"} []

kevinpapst commented:
Can you show me your mapping, maybe thats easier.

I wonder what this even is:
```
username: $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name
```
And please include a dump from src/Auth/User/SamlUserFactory.php / hydrateUser() function in line 50: `var_dump($token->getAttribues())`

timlegge commented:
The persist is actually the line causing the exception

timlegge commented:
```
kimai:
    saml:
        activate: true
        title: Login with Office 365
        mapping:
            - { saml: $Email, kimai: email }
            - { saml: $displayname, kimai: alias }
            - { saml: $title, kimai: title }
            - { saml: $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name, kimai: username }
```

timlegge commented:
So:

http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name is the nameid that SAML defines in its spec.  They can be overridden but it shows up in the saml response as the NameId:

     `<NameID Format="urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress">tim@example.com</NameID>`

timlegge commented:
Sorry I grabbed the wrong one:

```
      <Attribute Name="http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name">
        <AttributeValue>tim@example.com</AttributeValue>
      </Attribute>

```

kevinpapst commented:
And you really want to have a different username and email?
Can you try to remove the username mapping completely, test again.
I will setup a test for me with a different username. I always tested with email = username.

timlegge commented:
I can take a look at the configuration of our SAML response.  The username and email address are different for sure but I may be able to change it for this application.

timlegge commented:
I changed my saml to return the email address.  I deleted the user and logged in.  Next login same result - exception on login page no logs

Commented out the try and its trying to insert the duplicate using email address for the username:

I then commented out

//        try {
            $user = $this->userProvider->loadUserByUsername($token->getUsername());
//        } catch (UsernameNotFoundException $e) {
//        }

in  src/Auth/Provider/SamlProvider.php

This now shows that loadUserByUsername is not finding the user  although $token->getUsername() does fine my username:

string(26) "timothy.legge@example.com"

kevinpapst commented:
Ok, I could reproduce it with an invalid configuration, when SamlToken->getUsername() returns something different then the attributes "email" and "username".

If that is the case the user can't be found, because $token_>getUsername() is used to find the user on the next login. 

Conclusion: either email or username must match the token->getUsername().
I don't exactly know where/how $token->getUsername() is calculated, in G Suite it is my login email address. 

That why my config looks like this:
```
            - { saml: $Email, kimai: email }
            - { saml: $Email, kimai: username }
```

timlegge commented:
Odd.  I changed my mapping:

```
        mapping:
            - { saml: $Email, kimai: email }
            - { saml: $displayname, kimai: alias }
            - { saml: $title, kimai: title }
            - { saml: $Email, kimai: username }
```
 bin/console --env=dev cache:clear

databaes shows:

```
 select concat("'", username, "'") as username from kimai2_users;
+--------------------------------+
| username                       |
+--------------------------------+
| 'superadmin'                   |
| 'tim@googleacc.com' |
| 'timothy.legge@example.com'         |
+--------------------------------+
```
I did the concat to ensure there was no white space issues

and:
` var_dump($token->getUsername());exit;
`
returns:

`string(26) "timothy.legge@example.com" `

kevinpapst commented:
I am not sure that this will help ... but can you please retry with the latest push?

timlegge commented:
shortly

On Mon., Jan. 27, 2020, 12:51 p.m. Kevin Papst, <notifications@github.com>
wrote:

> I am not sure that this will help ... but can you please retry with the
> latest push?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kevinpapst/kimai2/pull/1408?email_source=notifications&email_token=AAH3N62Y2P34X2EBYVLGKLLQ74GIPA5CNFSM4KKZAKR2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKAG33Y#issuecomment-578842095>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH3N65EZ2YAYHKQ7OSZTTDQ74GIPANCNFSM4KKZAKRQ>
> .
>

timlegge commented:
Hi

I am at a loss.  Its still not working.

However I added:

var_dump($username);exit;

as line 118

of src/Repository/UserRepository.php

in function loadUserByUsername

Should that get called during the login because I am not seeing any output.

kevinpapst commented:
Yes it should be called when returning from your IDP at /auth/saml/acs. 
Okay, lets start from scratch with `git reset --hard` and `git pull`.
Remove the user from the database.
Cache clean is normally not necessary when running in dev mode, but lets do it nevertheless: `rm -rf var/cache/*`.
Then paste your local.yaml and remove the sensitive stuff.

timlegge commented:
Will do.  It looks like it is calling  the loadUserByUsername in src/Auth/User/DoctrineUserProvider.php

timlegge commented:
local.yaml sent to you via email in case I missed any thing

timlegge commented:
same behavour

timlegge commented:
Appears to fix my issue.  More testing


diff --git a/src/Auth/User/DoctrineUserProvider.php b/src/Auth/User/DoctrineUserProvider.php
index ed96e394..421725c1 100644
--- a/src/Auth/User/DoctrineUserProvider.php
+++ b/src/Auth/User/DoctrineUserProvider.php
@@ -42,7 +42,7 @@ final class DoctrineUserProvider implements UserProviderInterface

         // this needs improvements: should not check for LDAP user, but we have to ... as LDAP users cannot
         // be clearly identified by now, because the auth column was introduced after LDAP!
-        if (!$user->isInternalUser() && !$user->isLdapUser()) {
+        if (!$user->isInternalUser() && !$user->isLdapUser() && !$user->isSamlUser()) {
             throw new UsernameNotFoundException(sprintf('User "%s" is registered, but not as internal user.', $username));
         }

timlegge commented:
Issue was actually that I was missing a part of the config in local.yaml

Thanks @kevinpapst 
```
security:
    providers:
        chain_provider:
            chain:
                providers: [kimai_saml]
    firewalls:
        secured_area:
            saml:
                check_path: saml_acs
                login_path: saml_login
                failure_path: fos_user_security_login
```

timlegge commented on src/Auth/User/SamlUserFactory.php:
```diff
@@ -0,0 +1,111 @@
+            foreach ($this->groupMapping as $mapping) {
+                $field = $mapping['kimai'];
+                $attribute = $mapping['saml'];
+                $groupMap[$attribute] = $field;
```
Not a new feature request but is there a simple way to concatenate firstname and lastname that some IdP provide.  Something like:

        mapping:
            - { saml: $firstname . ' ' . $lastname, kimai: alias }

This might work now, I have not checked.  I don't need it for my installation

kevinpapst replied:
I thought about that as well, but I'll skip that for now as you don't need it ...

kevinpapst commented on src/Auth/User/DoctrineUserProvider.php:
```diff
@@ -0,0 +1,83 @@
+        }
+
+        // this needs improvements: should not check for LDAP user, but we have to ... as LDAP users cannot
+        // be clearly identified by now, because the auth column was introduced after LDAP!
```
rewritten => outdated

kevinpapst commented on src/Auth/Provider/SamlProvider.php:
```diff
@@ -0,0 +1,103 @@
+
+        try {
+            $user = $this->userProvider->loadUserByUsername($token->getUsername());
+        } catch (UsernameNotFoundException $e) {
```
rewritten/removed class. and in debug mode you see that the user is not found/created from the SQL statements

kevinpapst commented on tests/Auth/User/SamlUserFactoryTest.php:
```diff
@@ -0,0 +1,152 @@
+                ['saml' => 'A static super title', 'kimai' => 'title'],
+            ],
+            'roles' => [
+                'attribute' => 'RoLeS',
```
good eyes 😁 but the casing is wanted => its a test

kevinpapst commented:
Thanks for the review @timlegge !

I pushed a big update, to test:
- pull
- composer install
- change your config ([see docs](https://www.kimai.org/documentation/saml.html))
- flush cache

Some of the changes include: 
- The SamlUserProvider is removed (you need to remove your config!)
- SAML connection config was moved to kimai.saml.connection
- SAML users cannot reset their password
- Password for SAML users cannot be changed in their profile
- Manually registered users can login via SAML, but their account will be upgraded to **SAML only**

timlegge commented:
The role information is not updateing in the latest version.  Looking for the issue now

timlegge commented:
never mind.  My issue - reading the documentation caused me an issue:

roles:
            attribute: Roles

The attribute has to match the attribute name of the groups/roles being returned in the Saml Response

timlegge commented:
Found one issue with running behind a nginx proxy:

I will look at if thats a nginx resolution.  

[2020-01-29 16:06:40] request.CRITICAL: Uncaught PHP Exception RuntimeException: "The response was received at http://kimai-test.alc.ca:8001/auth/saml/acs instead of https://kimai-test.alc.ca/auth/saml/acs" at /opt/kimai/src/Saml/Controller/SamlController.php line 52 {"exception":"[object] (RuntimeException(code: 0): The response was received at http://kimai-test.alc.ca:8001/auth/saml/acs instead of https://kimai-test.alc.ca/auth/saml/acs at /opt/kimai/src/Saml/Controller/SamlController.php:52)"} []

kevinpapst commented:
My assumption is, that this is a general "problem" with the Symfony setup and not with the SAML module.
When you look at this: https://github.com/kevinpapst/kimai2/blob/master/public/index.php#L27 then you see, that the env variable `TRUSTED_PROXIES` tells Symfony to evaluate the `X-FORWARDED-*` headers and take them into account when generating URLs. 

Maybe you have to set the `TRUSTED_PROXIES` ENV.

How can I test that on my end? Can you share a snippet of your proxy config, so I can setup one here?

kevinpapst commented:
Ok, I did set it up and found a solution, which works at a first glance. Need more testing...

Here is my nginx proxy config (nothing special I would say):
```
server {
    listen 443 ssl;

    server_name kimai2.local;

    ssl_certificate      /Users/kevin/Sites/_certs/kimai2.local.crt;
    ssl_certificate_key  /Users/kevin/Sites/_certs/kimai2.local.key;
    ssl_session_timeout  5m;
    ssl_protocols  SSLv2 SSLv3 TLSv1;
    ssl_ciphers  HIGH:!aNULL:!MD5;
    ssl_prefer_server_ciphers   on;

    access_log  /Users/kevin/Sites/_logs/kimai2.local.log;

    location / {
          proxy_pass http://127.0.0.1:8010/;

          proxy_set_header  Host $http_host;
          proxy_set_header  X-Forwarded-Host $host:$server_port;
          proxy_set_header  X-Forwarded-Server $host;
          proxy_set_header  X-Real-IP $remote_addr;
          proxy_set_header  X-Forwarded-For $proxy_add_x_forwarded_for;
          proxy_set_header  X-Forwarded-Proto $scheme;
          proxy_set_header  X-Forwarded-Port $server_port;
      }
}
```

This is my .env setting:
```
TRUSTED_PROXIES=127.0.0.1,kimai2.local,localhost
```

And this is a saml specific part of my local.yaml (I think the `baseurl` is the important setting here):
```
            sp:
                entityId: 'https://kimai2.local/auth/saml/metadata'
                assertionConsumerService:
                    url: 'https://kimai2.local/auth/saml/acs'
                    binding: 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST'
                singleLogoutService:
                    url: 'https://kimai2.local/auth/saml/logout'
                    binding: 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect'
                #privateKey: ''
            # Optional settings
            baseurl: 'https://kimai2.local/auth/saml/'

```

timlegge commented:
Looking, my proxy is ... complex ...

kevinpapst commented:
Yes, but it is not about the proxy, but about the two kimai configs (the forwarded for headers are pretty standard, aren't they?).

kevinpapst commented:
@timlegge I found one more piece of code, which might help ... probably we can even skip the usage of the baseurl with that change. see my latest push, especially this line: https://github.com/kevinpapst/kimai2/pull/1408/commits/04223e09cad62d17294ef0332ed142ef8d27f701#diff-918f42d3001f48647a833788daea0e66R23

timlegge commented:
Hi

Its possible I have my proxies misconfigured but I still need the base enabled.  Works fine with that enabled.

Tim

kevinpapst commented:
LOL, for the other customer the baseurl didn't work, but this setting was required.
I have no plans for further changes, unless you find issues. So whenever you are done with testing. please let me know.

timlegge commented:
likely done tomorrow.

urinal-cake commented:
I had my team test SAML auth for about 3 hours yesterday, and we cannot find any issues as of today.

We did see an issue where the user would never be allowed to login if they first attempted to login using a non-SAML authenticated user:
1. Click Login button
2. Select Gmail account, or G Suite account for a different organization other than target.
3. Notice **app_not_configured_for_user** error
4. Go back in browser
5. Utilize G Suite account for target organization
6. Notice same error as #3, even though the account is now valid.

We're now no longer able to reproduce that issue, but will keep an eye out for it nonetheless. It could have been resolved by browser cache and Kimai app cache. However, we did that prior to starting our testing.

For the moment, I do agree with an earlier comment about the need to concatenate the 'first name' and 'last name' attribute. Google and idP do not provide 'Full Name' as an attribute. It is possible to create a 'Custom Attribute,' but for an organization with many users, this requires administrator access to do so, and can be time consuming.

It would be _incredibly_ powerful for us to be able to add the following:
`        mapping:
            - { saml: $Email, kimai: email }
            - { saml: $Email, kimai: username }
            - { saml: $FirstName ~ $LastName, kimai: alias }
`
Or something similar.

timlegge commented:
usually when I see app notconfigured is when I accidentally use my gmail.com
account

On Fri., Jan. 31, 2020, 11:34 a.m. urinal-cake, <notifications@github.com>
wrote:

> I had my team test SAML auth for about 3 hours yesterday, and we cannot
> find any issues as of today.
>
> We did see an issue where the user would never be allowed to login if they
> first attempted to login using a non-SAML authenticated user:
>
>    1. Click Login button
>    2. Select Gmail account, or G Suite account for a different
>    organization other than target.
>    3. Notice *app_not_configured_for_user* error
>    4. Go back in browser
>    5. Utilize G Suite account for target organization
>    6. Notice same error as #3
>    <https://github.com/kevinpapst/kimai2/issues/3>, even though the
>    account is now valid.
>
> We're now no longer able to reproduce that issue, but will keep an eye out
> for it nonetheless. It could have been resolved by browser cache and Kimai
> app cache. However, we did that prior to starting our testing.
>
> For the moment, I do agree with an earlier comment about the need to
> concatenate the 'first name' and 'last name' attribute. Google and idP do
> not provide 'Full Name' as an attribute. It is possible to create a 'Custom
> Attribute,' but for an organization with many users, this requires
> administrator access to do so, and can be time consuming.
>
> It would be *incredibly* powerful for us to be able to add the following:
> mapping: - { saml: $Email, kimai: email } - { saml: $Email, kimai:
> username } - { saml: $FirstName ~ $LastName, kimai: alias }
> Or something similar.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kevinpapst/kimai2/pull/1408?email_source=notifications&email_token=AAH3N6233EO7R2QRM2UDOKLRARAKFA5CNFSM4KKZAKR2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKPAODI#issuecomment-580781837>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH3N6Y2ELDOP7EENH46RHLRARAKFANCNFSM4KKZAKRQ>
> .
>

kevinpapst commented:
At that point in time kimai did not register the user, as no information about him is available - so I guess this problem won't reoccur.

I added the concatenation of attributes, configure it like this:
```
        mapping:
            - { saml: $Email, kimai: email }
            - { saml: $FirstName $LastName, kimai: alias }
```

urinal-cake commented:
Kevin, this worked flawlessly. We did some more testing with the combined fullname attribute, but haven't turned up any more issues. Good to go from our end. Let me know how else I can help.

kevinpapst commented:
@timlegge done as well or do you want some more testing time?

timlegge commented:
@kevinpapst have not pulled the latest commit.

kevinpapst commented:
Thanks @timlegge and @urinal-cake for all the testing!!!!

Changes can still be done, when we find any issue.
